File: sdks/go/apikeys.go
package fyredocs

import (
	"context"
	"net/http"
	"net/url"
)

// APIKeysAPI wraps the `/auth/api-keys/*` endpoints owned by
// auth-service.
type APIKeysAPI struct {
	c *Client
}

// ListAPIKeysOptions tunes a List call. Zero-value defaults are
// "active keys only".
type ListAPIKeysOptions struct {
	// Revoked, when true, switches to the audit archive of
	// revoked keys instead of active keys.
	Revoked bool
}

// List returns the calling user's API keys.
func (a *APIKeysAPI) List(ctx context.Context, opts *ListAPIKeysOptions) ([]APIKey, error) {
	q := url.Values{}
	if opts != nil && opts.Revoked {
		q.Set("revoked", "true")
	}
	var out []APIKey
	if err := a.c.Request(ctx, "/auth/api-keys", RequestOptions{
		Method: http.MethodGet,
		Query:  q,
		Out:    &out,
	}); err != nil {
		return nil, err
	}
	return out, nil
}

// Issue mints a new API key. The plaintext is in the returned
// response — display + persist it immediately, the server can't
// recover it.
func (a *APIKeysAPI) Issue(ctx context.Context, req IssueAPIKeyRequest) (*IssueAPIKeyResponse, error) {
	var out IssueAPIKeyResponse
	if err := a.c.Request(ctx, "/auth/api-keys", RequestOptions{
		Method: http.MethodPost,
		Body:   req,
		Out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

// Revoke marks the key with `id` as revoked. Idempotent — calling
// revoke on an already-revoked key is a no-op at the server.
func (a *APIKeysAPI) Revoke(ctx context.Context, id string) error {
	return a.c.Request(ctx, "/auth/api-keys/"+url.PathEscape(id)+"/revoke", RequestOptions{
		Method: http.MethodPost,
	})
}

File: sdks/python/src/fyredocs/_apikeys.py
"""``client.api_keys`` — auth-service /auth/api-keys/* wrapper."""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional
from urllib.parse import quote

from ._types import (
    APIKey,
    IssueAPIKeyRequest,
    IssueAPIKeyResponse,
    _api_key_from_json,
)

if TYPE_CHECKING:
    from ._client import Client


class APIKeysAPI:
    """Wraps ``/auth/api-keys/*``."""

    def __init__(self, client: "Client") -> None:
        self._c = client

    def list(self, *, revoked: bool = False) -> list[APIKey]:
        """List the calling user's API keys.

        Pass ``revoked=True`` to switch to the audit archive of
        revoked keys instead of active keys.
        """
        data = self._c.request(
            "/auth/api-keys",
            query={"revoked": "true"} if revoked else None,
        )
        return [_api_key_from_json(d) for d in (data or [])]

    def issue(self, req: IssueAPIKeyRequest) -> IssueAPIKeyResponse:
        """Mint a new API key. The plaintext is in the returned
        response — display + persist it immediately, the server
        can't recover it."""
        body: dict[str, object] = {"name": req.name}
        if req.environment:
            body["environment"] = req.environment
        if req.scopes is not None:
            body["scopes"] = list(req.scopes)
        data = self._c.request("/auth/api-keys", method="POST", body=body)
        if not isinstance(data, dict):
            raise ValueError("api_keys.issue: unexpected response shape")
        return IssueAPIKeyResponse(
            key=_api_key_from_json(data["key"]),
            plaintext=str(data.get("plaintext", "")),
        )

    def revoke(self, key_id: str) -> None:
        """Mark the key as revoked. Idempotent — calling revoke
        on an already-revoked key is a no-op at the server."""
        self._c.request(
            f"/auth/api-keys/{quote(key_id, safe='')}/revoke",
            method="POST",
        )
        return None
